x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/slm/SnapshotRetentionConfiguration.java

```diff
@@ -144,68 +146,68 @@ public Integer getMaximumSnapshotCount() {
+            if (this.minimumSnapshotCount != null && successfulSnapshotCount <= this.minimumSnapshotCount) {
+                if (UNSUCCESSFUL_STATES.contains(si.state()) == false) {
+                    logger.trace("[{}]: INELIGIBLE as there are {} non-failed snapshots ({} total) and {} minimum snapshots needed",
                             snapName, successfulSnapshotCount, totalSnapshotCount, this.minimumSnapshotCount);
-                        return false;
-                    } else {
-                        logger.trace("[{}]: SKIPPING minimum snapshot count check as this snapshot is {} and not counted " +
+                    return false;
+                } else {
+                    logger.trace("[{}]: SKIPPING minimum snapshot count check as this snapshot is {} and not counted " +
                             "towards the minimum snapshot count.", snapName, si.state());
-                    }
+                }
             }
 
             // Finally, check the expiration time of the snapshot, if it is past, then it is
             // eligible for deletion
             if (this.expireAfter != null) {
-                final TimeValue snapshotAge = new TimeValue(nowSupplier.getAsLong() - si.startTime());
-
                 if (this.minimumSnapshotCount != null) {
-                    final long eligibleForExpiration = Math.max(0, successfulSnapshotCount - minimumSnapshotCount);
-
                     // Only the oldest N snapshots are actually eligible, since if we went below this we
                     // would fall below the configured minimum number of snapshots to keep
-                    final Stream<SnapshotInfo> successfulSnapsEligibleForExpiration = sortedSnapshots.stream()
-                        .filter(snap -> SnapshotState.SUCCESS.equals(snap.state()))
-                        .limit(eligibleForExpiration);
-                    final Stream<SnapshotInfo> unsucessfulSnaps = sortedSnapshots.stream()
-                        .filter(snap -> unsuccessfulStates.contains(snap.state()));
-
-                    final Set<SnapshotInfo> snapsEligibleForExpiration = Stream
-                        .concat(successfulSnapsEligibleForExpiration, unsucessfulSnaps)
-                        .collect(Collectors.toSet());
-
-                    if (snapsEligibleForExpiration.contains(si) == false) {
+                    final boolean maybeEligible;
+                    if (si.state() == SnapshotState.SUCCESS) {
+                        maybeEligible = sortedSnapshots.stream()
+                            .filter(snap -> SnapshotState.SUCCESS.equals(snap.state()))
+                            .limit(Math.max(0, successfulSnapshotCount - minimumSnapshotCount))
+                            .anyMatch(si::equals);
+                    } else if (UNSUCCESSFUL_STATES.contains(si.state())) {
+                        maybeEligible = sortedSnapshots.contains(si);
+                    } else {
+                        maybeEligible = false;
```

Comment:
I think this branch should have its own `trace` log message and immediately return false, because at this point it is neither successful or unsuccessful, which means its state has prevented it from being considered (either `IN_PROGRESS` or `INCOMPATIBLE`). It would be confusing to log it as ineligible because of a minimum rather than ineligible due to its current state.